network-config-analyzer/CanonicalIntervalSet.py

```diff
@@ -230,27 +231,96 @@ def is_subset(self, other):
         :return: None
         """
-        new_interval_set = []
-        new_interval = interval_to_add.copy()  # so we don't change the original
-        new_interval_added = False
-        for interval in self.interval_set:
-            if not new_interval_added and (new_interval.overlaps(interval) or new_interval.touches(interval)):
-                # new_interval "swallows" existing interval (and will replace it)
-                new_interval = (interval | new_interval)[0]
-            else:
-                if not new_interval_added and new_interval.end < interval.start:
-                    new_interval_set.append(new_interval)
-                    new_interval_added = True
-                new_interval_set.append(interval)
-        if not new_interval_added:
-            new_interval_set.append(new_interval)
-        self.interval_set = new_interval_set
+        if not self:
+            self.interval_set.append(interval_to_add)
+            return
+        left = self.find_interval_left(interval_to_add)
+        right = self.find_interval_right(interval_to_add)
+
+        # interval_to_add has no overlapping/touching intervals between left to right
+        if left >= 0 and right >= 0 and right - left == 1:
+            self.interval_set.insert(left + 1, interval_to_add)
+            return
+
+        # interval_to_add has no overlapping/touching intervals and is smaller than first interval
+        if left == -1 and right == 0:
+            self.interval_set = [interval_to_add] + self.interval_set
+            return
+
+        # interval_to_add has no overlapping/touching intervals and is greater than last interval
+        if right == -1 and left == len(self.interval_set) - 1:
+            self.interval_set.append(interval_to_add)
+            return
+
+        # update left/right indexes to be the first potential overlapping/touching intervals from left/right
+        left += 1
+        right = right - 1 if right >= 0 else len(self.interval_set) - 1
+        # check which of left/right is overlapping/touching interval_to_add
+        left_overlaps = self.interval_set[left].overlaps(interval_to_add) or self.interval_set[left].touches(
+            interval_to_add)
+        right_overlaps = self.interval_set[right].overlaps(interval_to_add) or self.interval_set[right].touches(
+            interval_to_add)
+        # the interval_to_add has to be "merged" with overlapping/touching intervals
+        new_interval_start = min(interval_to_add.start,
+                                 self.interval_set[left].start) if left_overlaps else interval_to_add.start
+        new_interval_end = max(interval_to_add.end,
+                               self.interval_set[right].end) if right_overlaps else interval_to_add.end
+        new_interval = CanonicalIntervalSet.Interval(new_interval_start, new_interval_end)
+        self.interval_set = self.interval_set[:left] + [new_interval] + self.interval_set[right + 1:]
```

Comment:
I believe deleting `left`/`right` and then inserting `new_interval` might be even more efficient

Reply:
done